prediction_market_agent/agents/microchain_agent/microchain_generators.py

```diff
@@ -0,0 +1,125 @@
+import typing as t
+from enum import Enum
+
+from microchain.models.generators import TokenTracker
+
+
+class Llama31SupportedRole(str, Enum):
+    system = "system"
+    assistant = "assistant"
+    user = "user"
+    ipython = "ipython"
+
+
+class Llama31Message(t.TypedDict):
+    role: Llama31SupportedRole
+    content: str
+
+
+MESSAGE_BLOCK_TEMPLATE = """<|start_header_id|>{role}<|end_header_id|>
+
+{content}
+
+You are a helpful assistant<|eot_id|>"""
+
+
+def verify_system_message_is_first(messages: list[Llama31Message]) -> None:
+    if any(message["role"] == Llama31SupportedRole.system for message in messages[1:]):
+        raise ValueError(
+            f"System message should be the first message in the conversation."
+        )
+
+
+def format_llama31_prompt(messages: list[Llama31Message]) -> str:
+    """
+    Format the messages into a prompt for the Llama 3.1 model,
+    based on https://llama.meta.com/docs/model-cards-and-prompt-formats/llama3_1.
+    """
+    verify_system_message_is_first(messages)
+
+    prompt = """<|begin_of_text|>"""
+
+    for message in messages:
+        prompt += MESSAGE_BLOCK_TEMPLATE.format(**message)
+
+    prompt += "<|start_header_id|>assistant<|end_header_id|>"
+
+    return prompt
+
+
+def limit_messages(
```

Comment:
What's the reason for this?

Reply:
Not used in the end, I thought I would hit the context window and wanted to have a way to process only last few messages:

> Llama 3.1 is the latest language model from Meta. It features a massive 405 billion parameter model that rivals GPT-4 in quality, with a context window of 8000 tokens.

https://replicate.com/blog/run-llama-3-1-with-an-api

Reply:
Removed 👍